Sources/KeyboardShortcuts/Utilities.swift

```diff
@@ -559,3 +559,104 @@ extension Character {
 		self = Character(content)
 	}
 }
+
+final class UserDefaultsObservation: NSObject {
+	typealias Callback = (_ name: KeyboardShortcuts.Name, _ newKeyValue: String?) -> Void
+	
+	private let name: KeyboardShortcuts.Name
+	private let key: String
+	static var observationContext = 0
+	private weak var suite: UserDefaults?
+	private var isObserving = false
+	private let callback: Callback
+	private var lock = NSLock()
+	
+	init(
+		suite: UserDefaults,
+		name: KeyboardShortcuts.Name,
+		key: String,
+		_ callback: @escaping Callback
+	) {
+		self.suite = suite
+		self.name = name
+		self.key = key
+		self.callback = callback
+	}
+	
+	deinit {
+		invalidate()
+	}
+	
+	func start() {
+		lock.lock()
+		
+		guard !isObserving else {
+			return
+		}
+		
+		suite?.addObserver(
+			self,
+			forKeyPath: key,
+			options: [.new],
+			context: &Self.observationContext
+		)
+		isObserving = true
+		
+		lock.unlock()
```

Comment:
This is why most locking solution has a `withLock`. The code will deadlock if it's already observing as unlock is not called in the guard else.

```swift
extension NSLock {
	@discardableResult
	func withLock<T, E>(_ apply: () throws(E) -> T) throws(E) -> T {
		lock()

		defer {
			unlock()
		}

		return try apply()
	}
}
```

Reply:
Good catch, thanks. I'm using the `withLock` from the [Foundation library](https://github.com/swiftlang/swift-corelibs-foundation/pull/4736), if you don't mind.